src/workerd/api/headers.c++

```diff
@@ -0,0 +1,721 @@
+    KJ_CASE_ONEOF(idx, kj::uint) {
+      return commonHeaders[idx].map([](kj::Own<Header>& header) -> Header& { return *header; });
+    }
+    KJ_CASE_ONEOF(name, kj::String) {
+      return uncommonHeaders.find(name).map(
+          [](kj::Own<Header>& header) -> Header& { return *header; });
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+jsg::Ref<Headers> Headers::clone(jsg::Lock& js) const {
+  auto result = js.alloc<Headers>(js, *this);
+  result->guard = guard;
+  return kj::mv(result);
+}
+
+// Fill in the given HttpHeaders with these headers. Note that strings are inserted by
+// reference, so the output must be consumed immediately.
+void Headers::shallowCopyTo(kj::HttpHeaders& out) {
+  for (kj::uint i = 1; i < commonHeaders.size(); i++) {
+    KJ_IF_SOME(header, commonHeaders[i]) {
+      KJ_IF_SOME(name, header->name) {
+        for (auto& value: header->values) {
+          out.addPtrPtr(name, value);
+        }
+      } else {
+        auto name = getCommonHeaderName(i);
+        for (auto& value: header->values) {
+          out.addPtrPtr(name, value);
+        }
+      }
+    }
+  }
+
+  for (auto& header: uncommonHeaders) {
+    KJ_IF_SOME(name, header.value->name) {
+      for (auto& value: header.value->values) {
+        out.addPtrPtr(name, value);
+      }
+    } else {
+      for (auto& value: header.value->values) {
+        out.addPtrPtr(header.key, value);
+      }
+    }
+  }
+}
+
+kj::Array<Headers::DisplayedHeader> Headers::getDisplayedHeaders(jsg::Lock& js) {
+  auto getSetCookie = FeatureFlags::get(js).getHttpHeadersGetSetCookie();
+
+  size_t reserved = 0;
+
+  for (kj::uint i = 1; i < commonHeaders.size(); i++) {
+    KJ_IF_SOME(header, commonHeaders[i]) {
+      if (getSetCookie && i == static_cast<uint>(capnp::CommonHeaderName::SET_COOKIE)) {
+        reserved += header->values.size();
+      } else {
+        reserved += 1;
+      }
+    }
+  }
+  for (auto& header: uncommonHeaders) {
+    reserved += header.value->values.size();
+  }
+  kj::Vector<Headers::DisplayedHeader> vec(reserved);
+
+  for (kj::uint i = 1; i < commonHeaders.size(); i++) {
+    auto name = getCommonHeaderName(i);
+    KJ_IF_SOME(header, commonHeaders[i]) {
+      if (getSetCookie && i == static_cast<uint>(capnp::CommonHeaderName::SET_COOKIE)) {
+        for (auto& value: header->values) {
+          vec.add(Headers::DisplayedHeader{
+            .key = kj::str(name),
+            .value = kj::str(value),
+          });
+        }
+      } else {
+        vec.add(Headers::DisplayedHeader{
+          .key = kj::str(name),
+          .value = kj::strArray(header->values, ", "),
+        });
+      }
+    }
+  }
+
+  for (auto& header: uncommonHeaders) {
+    vec.add(Headers::DisplayedHeader{
+      .key = kj::str(header.key),
+      .value = kj::strArray(header.value->values, ", "),
+    });
+  }
+
+  auto ret = vec.releaseAsArray();
+  std::sort(ret.begin(), ret.end(), [](const auto& a, const auto& b) { return a.key < b.key; });
```

Comment:
Note that if we did at any point choose to support header key casing, this would need to be a cased sort.